storm-core/src/jvm/org/apache/storm/daemon/StormCommon.java

```diff
@@ -0,0 +1,605 @@
+                Map<String, Object> metricsConsumerConf = new HashMap<String, Object>();
+                metricsConsumerConf.put(Config.TOPOLOGY_TASKS, phintNum);
+                Bolt metricsConsumerBolt = Thrift.prepareSerializedBoltDetails(inputs, new MetricsConsumerBolt(className, argument), null, phintNum, metricsConsumerConf);
+
+                String id = className;
+                if (classOccurrencesMap.containsKey(className)) {
+                    // e.g. [\"a\", \"b\", \"a\"]) => [\"a\", \"b\", \"a#2\"]"
+                    int occurrenceNum = classOccurrencesMap.get(className);
+                    occurrenceNum++;
+                    classOccurrencesMap.put(className, occurrenceNum);
+                    id = Constants.METRICS_COMPONENT_ID_PREFIX + className + "#" + occurrenceNum;
+                } else {
+                    classOccurrencesMap.put(className, 1);
+                }
+                metricsConsumerBolts.put(id, metricsConsumerBolt);
+            }
+        }
+        return metricsConsumerBolts;
+    }
+
+    public static void addMetricComponents(Map conf, StormTopology topology) {
+        Map<String, Bolt> metricsConsumerBolts = metricsConsumerBoltSpecs(conf, topology);
+        for (Map.Entry<String, Bolt> entry : metricsConsumerBolts.entrySet()) {
+            topology.put_to_bolts(entry.getKey(), entry.getValue());
+        }
+    }
+
+    public static void addSystemComponents(Map conf, StormTopology topology) {
+        Map<String, StreamInfo> outputStreams = new HashMap<String, StreamInfo>();
+        outputStreams.put(Constants.SYSTEM_TICK_STREAM_ID, Thrift.outputFields(Arrays.asList("rate_secs")));
+        outputStreams.put(Constants.METRICS_TICK_STREAM_ID, Thrift.outputFields(Arrays.asList("interval")));
+        outputStreams.put(Constants.CREDENTIALS_CHANGED_STREAM_ID, Thrift.outputFields(Arrays.asList("creds")));
+
+        Map<String, Object> boltConf = new HashMap<String, Object>();
+        boltConf.put(Config.TOPOLOGY_TASKS, 0);
+
+        Bolt systemBoltSpec = Thrift.prepareSerializedBoltDetails(null, new SystemBolt(), outputStreams, 0, boltConf);
+        topology.put_to_bolts(Constants.SYSTEM_COMPONENT_ID, systemBoltSpec);
+    }
+
+    public static StormTopology systemTopology(Map stormConf, StormTopology topology) throws InvalidTopologyException {
+        return _instance.systemTopologyImpl(stormConf, topology);
+    }
+
+    protected StormTopology systemTopologyImpl(Map stormConf, StormTopology topology) throws InvalidTopologyException {
+        validateBasic(topology);
+
+        StormTopology ret = topology.deepCopy();
+        addAcker(stormConf, ret);
+        addEventLogger(stormConf, ret);
+        addMetricComponents(stormConf, ret);
+        addSystemComponents(stormConf, ret);
+        addMetricStreams(ret);
+        addSystemStreams(ret);
+
+        validateStructure(ret);
+
+        return ret;
+    }
+
+    public static boolean hasAckers(Map stormConf) {
```

Comment:
I prefer to just keep the same with origin code since these functions are related to "topology component", not a utils function. It is better to keep these kind functions in common.